proposals/pattern-variables.md

```diff
@@ -1,73 +1,96 @@
-    Use(x1);
-else if (e is (0, int x2))
-    Use(x2);
-  
-switch (e)
-{
-    case (int x, 0):
-        Use(x);
-        break;
-    case (0, int x):
-        Use(x);
-        break;
-}
-```
+- Allow variable declarations under `or` and `not` patterns and across `case` labels in a `switch` section.
+	```cs
+	if (e is C c or Wrapper { Prop: C c })
+	    return c;
+
+	Expr Simplify(Expr e)
+	{
+	  switch (e) {
+	    case Mult(Const(1), var x):
+	    case Mult(var x, Const(1)): 
+	    case Add(Const(0), var x):
+	    case Add(var x, Const(0)):
+		return Simplify(x);
+	    // ..
+	  }
+	}
+	```
+	Instead of:
+
+	```cs
+	if (e is C c1) 
+	    return c1;
+	if (e is Wrapper { Prop: C c2 }) 
+	    return c2;
 
+	Expr Simplify(Expr e)
+	{
+	  switch (e) {
+	    case Mult(Const(1), var x):
+		return Simplify(x);
+	    case Mult(var x, Const(1)): 
+		return Simplify(x);
+	    case Add(Const(0), var x):
+		return Simplify(x);
+	    case Add(var x, Const(0)):
+		return Simplify(x);
+	    // ..
+	  }
+	}
+	```
+
+- Also relax single-declaration rules across the entire variable scope as long as each variable is assigned once.
+	```cs
+	if (e is C c || e Wrapper { Prop: C c }) ;
+	if (e is C c) ; else if (e Wrapper { Prop: C c }) ;	
+	if (b ? e is C c : e is Warpper { Prop: C c }) ;
+	```
+	
 ## Detailed design
 
-Variables *must* be redeclared under all disjuncitve patterns because assignment of such variables depend on the order of evaluation which is undefined in the context of pattern-matching.
-
-- In a *disjunctive_pattern*, pattern variables declared on one side must be redeclared on the other side.
-- In a *switch_section*, pattern variables declared under each case label must be redeclared under every other case label.
-
-In any other case, variable declaration follows the usual scoping rules and is disallowed.
-
-These names can reference either of variables based on the result of the pattern-matching at runtime. Under the hood, it's the same local being assigned in each pattern.
-
-Redeclaring pattern variables is only permitted for variables of the same type.
-
-## Unresolved questions
-
-- How identical these types should be?
-- Could we support variable declarations under `not` patterns?
-    ```cs
-    if (e is not (int x, 0) and not (0, int x))
-    ```
-- Could we relax the scoping rules beyond pattern boundaries?
-    ```cs
-    if (e is (int x, 0) || a is (0, int x))
-    ```
-- Could we relax the redeclaration requirement in a switch section? 
-    ```cs
-    case (int x, 0) a when Use(x, a): // ok
-    case (0, int x) b when Use(x, b): // ok
-        Use(x); // ok
-        Use(a); // error; not definitely assigned
-        Use(b); // error; not definitely assigned
-        break;
-    ```
+### Variable redeclaration
+
+- Pattern variables are allowed to be redeclared within their scope if not already definitely assigned. These names can possibly reference either of variables based on the result of the pattern-matching at runtime.
```

Comment:
I think this should not include nested scoped, e.g.
```cs
if (e is C c) return;
if (e is C c) return; // allowed
{ if (e is C c) return; } // not allowed
```
Not completely sure if "within their scope" implies that.

Reply:
Perhaps `in the same scope`?

Reply:
It also would probably be worth moving it to an open questions section.

Reply:
There's a special error for that (CS0136) which I believe shouldn't be relaxed - in interaction with loops it gets too confusing. I think we should draw the line *somewhere* as the current rules are already too loose. Would like to see how all this play out.

Reply:
@333fred Done. Any chance this could be scheduled for ldm review? Thanks.